### DISCUSSION TITLE
Does supabase support database transactions?

### DISCUSSION THREAD

**Discussion by aditodkar** [supabase/supabase#4562] [Questions] [answered]
I have multiple tables and want to insert/update rows in these tables. I am using promise.all() inside nodejs API and each promise inside promise.all() is making insertion/updating data in database. Now lets say out of 5 promises 2 fail and 3 successful then in that case I don't want to update/insert data of those 3 successful promises. I want to update/insert all only if all are successful any idea how can this be done?

**Comment by GaryAustin1** [comment, marked as answer]
There have a been a couple of discussion on this: for example https://github.com/supabase/supabase/discussions/526
At the moment you have to use an rpc call to a postgres function that does your multiple database operations.

**Reply by korya** [reply]
Is it still the case in 2025?

**Reply by wzulfikar** [reply]
Yes :)
This is one common reason I see people use Drizzle and the like so they can do transactions (because direct access to db)

**Comment by geoguide** [comment]
I would like this please. `supabase.transaction(trx => ...)` would be awesome.

**Reply by Eigilak** [reply]
Amen brother!

**Reply by devjoseh** [reply]
🙏🙏🙏🙏 plss

**Reply by bill0x2A** [reply]
100% agree

**Reply by wzulfikar** [reply]
Having this while still retaining RLS would be nice. Until then, I guess I have to directly connect to the DB and use a raw SQL or orm to do transaction.

**Reply by maxwell-mutuku-dev** [reply]
Working on a project right now and it sucks having no logs for the supabase RPCs and working with objects in RPCs is a nightmare

**Reply by FlawaCLV** [reply]
+1

**Reply by alextbogdanov** [reply]
+1 Working with RPC has been a nightmare and I had to write my own server so that I can use Drizzle JUST to introduce transactions...

**Comment by drewbaker** [comment]
Our use case for transactions is for testing:

> One other random thing I've noticed about the way we're using Supabase (via their API) is that there's no way to use Postgres transactions.  The way I have often setup a unit test that depends on the database is to run the whole test inside a database transaction.  That way each test effectively gets it's own isolated database since at the end of the test you can just rollback the transaction without committing it.  This is not possible through the Supabase API so the tests just need to make sure any data they use is unique to that test and will not interfere with other tests.  Not a huge deal but the lack of transactions in general seems potentially worrisome.  The only way it's possible (as far as I can tell) is to write a custom SQL function that you can call via the Supabase API.  That's fine but means you can't do something like this in the code:
```
// pseudocode
await transaction(() => {
   const x = await supabase.from("some_table").select("x")...
   const y = someCalculation(x)
   await supabase.from("some_table").update({ y })...
});
```
> You'd have to implement `someCalculation` in SQL to have the full select + update be transactional.  For an application that will be dealing with financial stuff in the future I think this could become limiting.

**Comment by sunapi386** [comment]
It's now June 2025 and this is still is not supported, I'm in _awe_

**Comment by pedroresende** [comment]
And another year has passed and for transaction we need to rely on rpc 😢
